File: JavaScript/4-cache.js
'use strict';

function fn() {
  console.log('Generate cache');
  let cache = {};
  return function(a) {
    let res = cache[a];
    if (res) {
      console.log('From cache');
      return res;
    } else {
      console.log('Calculate and save to cache');
      res = 'value' + a;
      cache[a] = res;
      return res;
    }
  };
}

let f1 = fn();
let f2 = fn();
f1(1);
f1(2);
f1(1);
f1(2);
f2(1);
f2(2);
f2(1);
f2(2);

File: Python/4-cache.py
def fn():
    print('Generate cache')
    cache = {}
    def get_from_cache(key):
        res = cache.get(key)
        if res:
            print('From cache')
            return res
        else:
            print('Calculate and save')
            res = 'value ' + str(key)
            cache[key] = res
    return get_from_cache

f1 = fn()
f2 = fn()

f1(1)
f1(2)
f1(1)
f1(2)

f2(1)
f2(2)
f2(1)
f2(2)